mongo/change_stream.go

```diff
@@ -284,48 +283,65 @@ func (cs *ChangeStream) executeOperation(ctx context.Context, resuming bool) err
 		// Cancel the timeout-derived context at the end of executeOperation to avoid a context leak.
 		defer cancelFunc()
 	}
-	if original := cs.aggregate.Execute(ctx); original != nil {
-		retryableRead := cs.client.retryReads && cs.wireVersion != nil && cs.wireVersion.Max >= 6
-		if !retryableRead {
-			cs.err = replaceErrors(original)
-			return cs.err
+
+	// Execute the aggregate, retrying on retryable errors once (1) if retryable reads are enabled and
+	// infinitely (-1) if context is a Timeout context.
+	var retries int
+	if cs.client.retryReads && cs.wireVersion != nil && cs.wireVersion.Max >= 6 {
+		retries = 1
+	}
+	if internal.IsTimeoutContext(ctx) {
+		retries = -1
+	}
+
+	var err error
+AggregateExecuteLoop:
+	for {
```

Comment:
How does this retry loop work when called in the retry loop in [`ChangeStream.loopNext`](https://github.com/mongodb/mongo-go-driver/blob/1720e221d95d793edd1d0193cf3813bcd21fabec/mongo/change_stream.go#L624)? Are both necessary?

Reply:
Great question! The call to `executeOperation` in `loopNext` is reached when the `getMore` from a `Next` call encounters a resumable error. In other words, that call to `executeOperation` is used to "resume" the change stream: attempt to re-initialize the change stream with the original options. If the `Client` used to create the `ChangeStream` had a `Timeout` specified, then that resumption process will retry as many times as possible. While not specified in the CSOT spec explicitly, this seems like correct behavior to me.

The `for` in `ChangeStream.loopNext` still seems necessary to me, because even without specifying `Timeout`, if a user calls `Next` we want to retry `getMore` until a non-empty batch is returned or an error occurs. That retried `getMore` is distinct from the retry of the original `aggregate` or a resuming `aggregate`.

Reply:
Makes sense, thanks for the explanation!